### ISSUE TITLE
internal error when using named argument

### ISSUE THREAD

**Issue by jhh67** [chapel-lang/chapel#20373]
### Summary of Problem
I got the following error
```
examples/baz.chpl:12: In method 'encode':
examples/baz.chpl:13: internal error: RES-ADD-LLS-0805 chpl version 1.28.0 pre-release (0085fba89a)

Internal errors indicate a bug in the Chapel compiler ("It's us, not you"),
and we're sorry for the hassle.  We would appreciate your reporting this bug --
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.

  examples/baz.chpl:18: called as Foo.encode(key: nothing, value: int(64))
note: generic instantiations are underlined in the above callstack
```

### Steps to Reproduce

**Source Code:**

```chapel
class Encoder {
  // Encodes something, with optional key
  proc encode(ch:Foo, key:?t = none, x) : void throws {writeln("base");}
}

class DefaultEncoder: Encoder {
  override proc encode(ch:Foo, key:?t = none, x) : void throws { writeln("default");}
}

record Foo {
    var encoder: owned Encoder = new DefaultEncoder();
    proc encode(key:?t = none, value) {
        this.encoder.encode(this, key, value);
    }
}

var foo = new Foo();
foo.encode(value=42);
```

**Compile command:**
`chpl baz.chpl`

**Execution command:**
`./baz`

**Associated Future Test(s):**

### Configuration Information

- Output of `chpl --version`:
```
chpl version 1.28.0 pre-release (0085fba89a)
Copyright 2020-2022 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: clang
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: none *
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: none *
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```
- (For Cray systems only) Output of `module list`:

**Comment by bradcray** [comment]
Strangely, the developer view of that error looks like it could almost be a user error, except that the counts reflect internal arguments that a user wouldn't be aware of:

```
testit.chpl:12: In method 'encode':
testit.chpl:13: internal error: number of actuals (5) does not match number of formals (4) in encode() [resolution/addAutoDestroyCalls.cpp:805]
  testit.chpl:18: called as Foo.encode(key: nothing, value: int(64))
note: generic instantiations are underlined in the above callstack
```

**Comment by mppf** [comment]
I looked at this a bit in the debugger & I think that the problem is that the `none` formal got removed from the formals but not from the call at this point in resolution.

**Comment by bradcray** [comment]
Belated realization:  In:

```chapel
    proc encode(key:?t = none, value) {
```

the default value of `none` suggests that the only legal type for `key: ?t` is `nothing` which suggests that that first argument isn't actually very useful, whether using the default or passing `none` in explicitly.  Of course, we still shouldn't generate an internal error for it, but just noting that this isn't giving a completely generic option type as a programmer coming with a language that supported them might think—right?

Tagging @daviditen on this who I believe implemented these features (and may know something about the removal of the arguments, if it's at all related to our removal of `void` types and values (?).

**Comment by mppf** [comment]
> the default value of `none` suggests that the only legal type for `key: ?t` is `nothing` which suggests that that first argument isn't actually very useful, whether using the default or passing `none` in explicitly.

The point of it is that you can pass in something other than `none` and it will accept it. That is what we get out of writing `key:?t`. We would not get that behavior if the `?t` is dropped and it is just `key=none`.

**Comment by bradcray** [comment]
Oh gosh, you're right!  I was under the mistaken impression that `formal: ?t = defaultExpr` required `t` to be a type that could legally be assigned a `defaultExpr`.  But maybe my brain is just stuck in prehistoric times here and failed to digest a more recent change?  [that's your cue to embarrass me further by telling me how long ago that change occurred]

Did you happen to ascertain why the `none` formal got removed?  I'm specifically wondering whether it is related to `void`/`none`/`nothing` being eliminated from the compiler as an optimization or not.  Replacing `none` with `42` seems to cause the error to go away, suggesting it probably is (?).  In which case, @daviditen might be a very logical owner for fixing this.

**Comment by bradcray** [comment]
(I edited the issue title to try and make it a bit more specific to the use of `none`... assuming I didn't misdiagnose anything)

**Comment by mppf** [comment]
> that's your cue to embarrass me further by telling me how long ago that change occurred

Definitely not a goal here. Of course any of us could try some old releases to see. I thought it was always this way though...

> Did you happen to ascertain why the none formal got removed?

no

> I'm specifically wondering whether it is related to void/none/nothing being eliminated from the compiler as an optimization or not.

I think it is related to that logic, but I haven't dug in further to know for sure.

**Comment by bradcray** [comment]
> Definitely not a goal here.

I wasn't implying that was your goal, just that it was the way I'd feel if we had discussed and changed it at some point and I'd forgotten.  :)  I haven't bothered to go back and try to find out.

@daviditen:  Still hoping from a word for you as to whether you agree that the void/none removal pass is likely to be the issue here, and if so whether it might have a simple fix.

**Comment by daviditen** [comment]
Yes, it looks like the `nothing` formal is being removed but the `none` value being passed isn't. I'm not positive what about this case is different from the cases that are working correctly, but the named argument and the generic on the none formal are likely culprits.
